Overhaul Object Store Backends

jmchilton opened the pull request:
- Add a boto3 object store with tests (if you have keys defined) and update documentation. It works well for me both against AWS and against other S3 via interop (I used GCP another clear target of interest for the project). This is the first time Galaxy has had test cases for the multithreaded transfer to S3 object stores.
- I removed a ton of code (mostly duplicated code that was merged to a parent class but also some code that didn't make sense or wasn't used) from all the non-disk object stores. Over 400 lines of code removed each from S3, cloud, and Azure object stores and hundreds more removed from pithos, rucio, and irods.
- I think the Azure and the Boto3 object stores are now pretty good and clean examples of how to implement a caching object store - pull all the caching logic from the base class and just expose small methods that are very precise and efficient in terms of how they access cloud APIs, respond to error, and such. S3's legacy is on display in the implementation and I didn't break any of that (in fact there are more tests than ever) but it isn't a great example as a result. On the other hand the cloudbridge implementation is very broad with exception handling and that isn't great. Also I think in trying to be so generic you miss a lot of options available through boto3.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

jmchilton committed:
Add test case for GCP S3 Interop Object Stores.

Works with service account and user accounts keys. Settings -> Interoperability to create HMAC keys.

jmchilton committed:
De-dupclication around object store permission restoration.

jmchilton committed:
De-duplication around object store _construct_path.

jmchilton committed:
Reduce duplication around _pull_into_cache.

jmchilton committed:
Converge a few cloud object stores toward same names ahead of refactor.

jmchilton committed:
De-duplicate object store _exists across 4 object stores.

jmchilton committed:
Remove unused object store method.

jmchilton committed:
Remove duplicated _create logic across 4 object stores.

jmchilton committed:
Remove duplicated _empty across caching object stores.

jmchilton committed:
Refactor names to make clear copy-paste.

jmchilton committed:
Remove duplication around _get_data in object stores.

jmchilton committed:
Remove _in_cache from s3 (rebase above somewhere?)

jmchilton committed:
Remove duplication around _get_filename.

jmchilton committed:
Refactor object stores so we can remove _delete...

jmchilton committed:
Remove duplication around _delete in object stores.

jmchilton committed:
Implement a boto3 object store.

jmchilton committed:
Eliminate copy-paste for _size across 4 objct stores.

jmchilton committed:
Remove a bunch of extra imports now.

jmchilton committed:
Remove some duplication around cache targets.

jmchilton committed:
Remove unused copy-pasta from cloud object store.

jmchilton committed:
Remove wrong comments.

jmchilton committed:
Remove duplication around cache monitor starting.

jmchilton committed:
Converge _download a bit if needed.

jmchilton committed:
Re-work object store de-duplication for proper typing.

jmchilton committed:
Fix duplication around cache logging.

Eliminate CloudConfigMixin...

This got copid and pasted over and over and each place just used once. Really odd reading of my original code but I guess I understand why people did it.

jmchilton committed:
De-duplicat _push_to_os.

jmchilton committed:
Fix axel call across s3 & cloud object stores.

jmchilton committed:
Merge axel download stuff together.

jmchilton committed:
Remove get_file_name from cloud.

jmchilton committed:
Remove duplication around _update_from_file.

jmchilton committed:
Implement advanced transfer options for boto3.

jmchilton committed:
Fix irods test cases for code refactoring.

jmchilton committed:
Rev object store docs.

jmchilton committed:
Allow older style connection parameters on newer boto3 object store.

Should make migration for admins a bit easier.

jmchilton committed:
Drop logging that contains a password.

mvdbeek commented on lib/galaxy/objectstore/_caching_base.py:
```diff
@@ -0,0 +1,384 @@
+        # Now pull in the file
+        file_ok = self._download(rel_path)
+        fix_permissions(self.config, self._get_cache_path(rel_path_dir))
+        log.debug("_pull_into_cache: %s\n\n\n\n\n\n", ipt_timer)
```
That might have been there before, but can we drop at least the extra newlines here ?

jmchilton replied:
Yeah - debugging I left in - I'll drop the whole thing.

mvdbeek commented on lib/galaxy/objectstore/_caching_base.py:
```diff
@@ -0,0 +1,384 @@
+
+        # TODO: Sync should probably not be done here. Add this to an async upload stack?
+        if in_cache and not exists_remotely:
+            self._push_to_os(rel_path, source_file=self._get_cache_path(rel_path))
```
Is it time to also rename `_push_to_os` to maybe `_push_to_storage` or `_push_to_remote` ? I always get confused by that

jmchilton replied:
Yes!

mvdbeek commented on lib/galaxy/objectstore/_caching_base.py:
```diff
@@ -0,0 +1,384 @@
+
+    def _in_cache(self, rel_path: str) -> bool:
+        """Check if the given dataset is in the local cache and return True if so."""
+        cache_path = self._get_cache_path(rel_path)
```
There are open bugs (https://github.com/galaxyproject/galaxy/issues/15283, maybe https://github.com/galaxyproject/galaxy/issues/16438), where the cached copy is of size 0 while the remote file is complete. Should we check the size ?

jmchilton replied:
Comment in ``_get_filename``:

```python
        # Check if the file exists in the cache first, always pull if file size in cache is zero
        if self._in_cache(rel_path) and (dir_only or os.path.getsize(self._get_cache_path(rel_path)) > 0):
            return cache_path
```

I feel like I vaguely prefer doing it a layer above ``_in_cache`` like that. And it seems like it has the behavior you're looking for as long as ``sync_cache`` is un-sent as an argument or explicitly passed as True.

So I am anxious this is not the right approach and results in more API calls. For https://github.com/galaxyproject/galaxy/issues/15283 - hmm... maybe we make:

```
        file_ok = self._download(rel_path)
        fix_permissions(self.config, self._get_cache_path(rel_path_dir))
        return file_ok
```

in _pull_into_cache into:

```
        file_ok = self._download(rel_path)
        if file_ok:
             fix_permissions(self.config, self._get_cache_path(rel_path_dir))
        else:
              unlink(self._get_cache_path(rel_path), ignore_errors=True)
        return file_ok
```

This would seem to prevent an important class of errors. I could see it fixing https://github.com/galaxyproject/galaxy/issues/15283  but I doubt it is the problem with https://github.com/galaxyproject/galaxy/issues/16438. I think that problem would be better address by bailing on object store creation if a target job cache directory is not writable? I know this isn't a perfect fix for https://github.com/galaxyproject/galaxy/issues/16438 but it wouldn't leave us in an inconsistent state where we're having to deduce why things didn't work and wonder what exception handler is broken.

Do you have any objections to these changes as an effort to address those two issues?

mvdbeek replied:
No objections, sounds like a solid improvement, thank you!

jmchilton replied:
And look we will be fixing 6 object stores at once instead of making changes to copy and pasted code shared between all of them but only fixed in one 😆

mvdbeek reviewed (commented):
Looks great!

jmchilton committed:
Revise object store changes based on PR review.

jmchilton committed:
Prevent incomplete files from sticking around in the object store cache.

Maybe fixes https://github.com/galaxyproject/galaxy/pull/18117.

jmchilton committed:
Do not allow creation of object stores with non-writable caches.

Fixes biggest issue with #16438 - IMO anyway.

mvdbeek commented on lib/galaxy/objectstore/_caching_base.py:
```diff
@@ -0,0 +1,396 @@
+                raise Exception(f"Caching object store created with path '{staging_path}' that does not exist")
+
+        if not os.access(staging_path, os.R_OK):
+            raise Exception(f"Caching object store created with path '{staging_path}' that does not readable")
+        if not os.access(staging_path, os.W_OK):
+            raise Exception(f"Caching object store created with path '{staging_path}' that does not writable")
```
```suggestion
            raise Exception(f"Caching object store created with path '{staging_path}' that is not readable")
        if not os.access(staging_path, os.W_OK):
            raise Exception(f"Caching object store created with path '{staging_path}' that is not writable")
```

jmchilton committed:
Various object store extra files handling fixes.

- Fix s3 which broke during the migration because we stopped adding a / I thought was silly.
- Use the test case I wrote that mimics the integration test failure to also fix the boto3 and azure_blob object stores for these failures.

Also make note of a lingering issue that is pretty important

jmchilton committed:
Fix extra files handling for cached object stores when reset_cache called.
